src/ds++/test/cxx11example_move_semantics.cc

```diff
@@ -0,0 +1,202 @@
+ *
+ * \sa http://blog.smartbear.com/c-plus-plus/c11-tutorial-introducing-the-move-constructor-and-the-move-assignment-operator/
+ */
+//---------------------------------------------------------------------------//
+
+#include "ds++/Release.hh"
+#include "ds++/ScalarUnitTest.hh"
+#include "ds++/Soft_Equivalence.hh"
+#include <iostream>
+#include <iterator>
+#include <string>
+#include <vector>
+
+// forward declaration
+void report_memory_locations(std::vector<double> const &v,
+                             std::string const &name);
+
+// Classes used by the example
+
+//============================================================================//
+/*!
+ * \class A
+ * \brief Improper use of move semantics in constructor.
+ */
+//============================================================================//
+
+struct A {
+
+public:
+  /*!
+   * \brief Constructor
+   * \param[in] v_in v_in is lvalue; its type is rval ref
+   */
+  explicit A(std::vector<double> &&v_in)
+      : v_(v_in) // regular std::vector copy ctor called b/c lvalue
+  {
+    /* empty */
+  }
+
+  // member data
+  std::vector<double> v_;
+};
+
+//============================================================================//
+/*!
+ * \class B
+ * \brief Proper use of move semantics in constructor.
+ */
+//============================================================================//
+
+struct B {
+
+  /*!
+   * \brief Constructor
+   * \param[in] v_in v_in is lvalue; its type is rval ref
+   */
+  explicit B(std::vector<double> &&v_in)
+      : v_(std::move(v_in)) /* std::move casts to rval, move ctor called */
+  {
+    /* empty */
+  }
+
+  // member data
+  std::vector<double> v_;
+};
+
+//----------------------------------------------------------------------------//
+/*! \breif Demonstration of move semantics
+ *
+ * 1. Create a vector
+ * 2. Attempt to construct a class (A), demonstrate that ownership is not
+ *    transferred.
+ * 3. Attempt to construct another class (B), demonstrate that ownership is
+ *    transferred.
+ */
+//----------------------------------------------------------------------------//
+void move_semantics_example(rtt_dsxx::UnitTest &ut) {
+  using namespace std;
+
+  cout << ">>> Begin demonstration...\n" << endl;
+
+  // Create a complex object.
+  cout << "Create a vector v1.\n";
+  vector<double> v1 = {1, 2, 3};
+  report_memory_locations(v1, "v1");
+
+  auto const v1_loc = &v1;
+  auto v1_data_loc = &v1[0];
+
+  // Case 1:
+  // Create an object, attempt to transfer ownership from v1 to a.
+  // This will fail.
+  cout << "\nCreate an instantiation of A that owns a copy of v1.";
+  A a(move(v1));
+  cout << "\nAfter call to A::ctor\n";
+  report_memory_locations(v1, "v1");
+  report_memory_locations(a.v_, "a.v_");
+
+  // v1 remains unchanged! (not the behavior we want)
+  if (v1_loc != &v1)
+    ITFAILS;
+  if (v1_data_loc != &v1[0])
+    ITFAILS;
+  else
+    PASSMSG("v1 remains unchanged! (but this is not the behavior we want).");
```

Comment:
Could the parenthetical remark be changed to something like, "(but v1 should be uninitialized)"?